### PR TITLE
API Entries: implemented all but list and wait

### PR DESCRIPTION
**Pull request by MilkFansHelloWorld**
Implemented API entries
Created structs for request/response
Created a simple validation function and error handling
Made Connect4Adapter public so accessible by API
Made adapter and connect4 modules public
Created ::from methods for both Id structs
Uncommented placeholders for GameManager functions

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -1,33 +1,151 @@
...
+#[derive(Debug, Clone)]
+pub enum ValidationError {
+    ParseIntError(String),
+    NoSuchGameError(String)
+}
+
+impl Display for ValidationError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ValidationError::ParseIntError(id) => write!(f, "ID {} is Invalid", id),
```
Comment (konsolas) [inline src/api.rs:28]: Should this be called a ParseIDError in case we change the format of IDs?

File: src/api.rs [hunk 2]
```diff
@@ -1,33 +1,151 @@
...
-use actix_web::{post, get, Responder, web::Payload};
+use std::fmt::{Display, Formatter, write};
+use std::ops::DerefMut;
+use std::str::FromStr;
+use std::sync::RwLock;
+use actix_web::{post, get, Responder, web, Result, ResponseError, Error};
+use actix_web::http::StatusCode;
+use actix_web::web::Json;
+use serde::{Serialize, Deserialize};
+use serde_json::{json};
```
Comment (konsolas) [inline src/api.rs:11]: Braces not needed - run `cargo fmt`?

File: src/api.rs [hunk 3]
```diff
@@ -1,33 +1,151 @@
...
 
 #[post("/api/{game_id}/submit-move")]
-pub(crate) async fn submit_move(mut payload: Payload) -> impl Responder {
-    todo!(); "not yet implemented"
+pub(crate) async fn submit_move(web::Path((game_id)): web::Path<String>,
+                                mut payload: web::Json<adapter::GenericGameMove>,
+                                gm_wrapped: web::Data<RwLock<GameManager>>) -> Result<Json<SubmitMoveRes>> {
+    let game_id = validate_id(&game_id)?;
+    // Submit that to game manager
+    let mut gm = gm_wrapped.write().unwrap();
```
Comment (konsolas) [inline src/api.rs:143]: We shouldn't need a write lock to submit a move as the game adapter map uses interior mutability through `Mutex`

File: src/api.rs [hunk 4]
```diff
@@ -1,33 +1,151 @@
...
+        Ok(value) => Ok(value),
+        Err(_) => return Err(actix_web::Error::from(ValidationError::ParseIntError(game_id.clone()))) // TODO: need custom error
+    }
+}
+
+/* Structs for the different endpoints*/
+
+/// Struct for req to `/api/create-game`
+#[derive(Deserialize)]
+pub struct CreateGameReq {
```
Comment (konsolas) [inline src/api.rs:53]: Do these all need to be `pub`? Are they used anywhere else?

File: src/api.rs [hunk 5]
```diff
@@ -1,33 +1,151 @@
...
 #[post("/api/create-game")]
-pub(crate) async fn create_game(mut payload: Payload) -> impl Responder {
-    todo!(); "not yet implemented"
+pub(crate) async fn create_game(mut payload: web::Json<CreateGameReq>,
+                                gm_wrapped: web::Data<RwLock<GameManager>>) -> Result<Json<CreateGameRes>> {
+    // Validate & Pass it on
+    let mut gm = gm_wrapped.write().unwrap();
+    let game_id = match payload.name.as_str() {
+        "connect_4" => gm.create_game(|id| Box::new(connect4::Connect4Adapter::new(id))),
+        _ => return Err(Error::from(NoSuchGameError(payload.name.clone()))) // TODO: Should I return an empty string or an error
```
Comment (konsolas) [inline src/api.rs:107]: Error is good

File: src/api.rs [hunk 6]
```diff
@@ -1,33 +1,151 @@
...
-pub(crate) async fn submit_move(mut payload: Payload) -> impl Responder {
-    todo!(); "not yet implemented"
+pub(crate) async fn submit_move(web::Path((game_id)): web::Path<String>,
+                                mut payload: web::Json<adapter::GenericGameMove>,
+                                gm_wrapped: web::Data<RwLock<GameManager>>) -> Result<Json<SubmitMoveRes>> {
+    let game_id = validate_id(&game_id)?;
+    // Submit that to game manager
+    let mut gm = gm_wrapped.write().unwrap();
+    // Return success or failure
+    gm.receive_move(payload.into_inner()).and_then(|()|Ok(Json(SubmitMoveRes { success: true })))
```
Comment (konsolas) [inline src/api.rs:145]: `cargo fmt` pls

File: src/game/mod.rs [hunk 7]
```diff
@@ -24,6 +26,10 @@ impl GameId {
         let id: u128 = rng.gen();
         GameId(id)
     }
+    // Added for API to create a GameId object to input to the GameManager
+    pub fn from(id: u128) -> Self {
```
Comment (konsolas) [inline src/game/mod.rs:30]: Is `GameId::from(id)` really better than `GameId(id)`?

**Review by konsolas — CHANGES_REQUESTED** [review]

File: src/api.rs (hunk 6 above)
Comment (MilkFansHelloWorld) [reply src/api.rs:145]: Ok

File: src/api.rs (hunk 5 above)
Comment (MilkFansHelloWorld) [reply src/api.rs:107]: Ok

File: src/api.rs (hunk 4 above)
Comment (MilkFansHelloWorld) [reply src/api.rs:53]: Ok

File: src/api.rs (hunk 3 above)
Comment (MilkFansHelloWorld) [reply src/api.rs:143]: Ok

File: src/api.rs (hunk 2 above)
Comment (MilkFansHelloWorld) [reply src/api.rs:11]: Ok

File: src/api.rs (hunk 1 above)
Comment (MilkFansHelloWorld) [reply src/api.rs:28]: Ok

File: src/game/mod.rs (hunk 7 above)
Comment (MilkFansHelloWorld) [reply src/game/mod.rs:30]: Will put it in the game module

File: src/api.rs (hunk 4 above)
Comment (MilkFansHelloWorld) [reply src/api.rs:53]: Yes they do it turns out.

File: src/game/mod.rs [hunk 8]
```diff
@@ -38,6 +80,10 @@ impl SessionId {
         let id: u128 = rng.gen();
         SessionId(id)
     }
+    // Added for API to create a GameId object to input to the GameManager
+    pub fn from(id: u128) -> Self {
```
Comment (konsolas) [inline src/game/mod.rs:84]: Should this be consistent with `GameId::from` (i.e. take a &String)?

File: src/api.rs [hunk 9]
```diff
@@ -1,33 +1,136 @@
...
+) -> Result<Json<CreateGameRes>> {
+    // Validate & Pass it on
+    let mut gm = gm_wrapped.write().unwrap();
+    let game_id = match payload.name.as_str() {
+        "connect_4" => gm.create_game(|id| Box::new(connect4::Connect4Adapter::new(id))),
+        _ => return Err(Error::from(NoSuchGameError(payload.name.clone()))),
+    };
+    game_id.and_then(|id| {
+        Ok(Json(CreateGameRes {
+            game_id: id.to_string()[4..].parse()?,
```
Comment (konsolas) [inline src/api.rs:82]: This is very suspicious... if the `Display` implementation for `GameId` returns something like "Game12345", why do we strip the "Game" here before sending it to the client? Wouldn't it be helpful when debugging for a user to know at first glance what sort of id they have?

File: src/game/mod.rs [hunk 10]
```diff
@@ -24,6 +49,23 @@ impl GameId {
         let id: u128 = rng.gen();
         GameId(id)
     }
+    // Added for API to create a GameId object to input to the GameManager
+    pub fn from(id: &String) -> Result<Self> {
```
Comment (konsolas) [inline src/game/mod.rs:53]: Probably need a discussion about whether or not this should be consistent with `impl Display for GameId` (i.e. with the prefix or not)

**Review by konsolas — APPROVED** [review]
LGTM

**Review by jonathanjameswatson — APPROVED** [review]
All looks good

**Pull request merged**